src/main/java/io/curity/identityserver/plugin/OptInMFA/OptInMFAuthenticationActionHandler.java

```diff
@@ -0,0 +1,114 @@
+ */
+package io.curity.identityserver.plugin.OptInMFA;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import se.curity.identityserver.sdk.NonEmptyList;
+import se.curity.identityserver.sdk.attribute.Attribute;
+import se.curity.identityserver.sdk.authenticationaction.completions.ActionCompletionRequestHandler;
+import se.curity.identityserver.sdk.authenticationaction.completions.ActionCompletionResult;
+import se.curity.identityserver.sdk.errors.AuthenticatorNotConfiguredException;
+import se.curity.identityserver.sdk.service.ExceptionFactory;
+import se.curity.identityserver.sdk.service.SessionManager;
+import se.curity.identityserver.sdk.service.authenticationaction.AuthenticatorDescriptor;
+import se.curity.identityserver.sdk.service.authenticationaction.AuthenticatorDescriptorFactory;
+import se.curity.identityserver.sdk.web.Request;
+import se.curity.identityserver.sdk.web.Response;
+import se.curity.identityserver.sdk.web.cookie.Cookie;
+
+import java.lang.invoke.MethodHandles;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.curity.identityserver.plugin.OptInMFA.OptInMFAAuthenticationAction.AVAILABLE_SECOND_FACTORS_ATTRIBUTE;
+import static io.curity.identityserver.plugin.OptInMFA.AuthenticatorModel.toAuthenticatorModel;
+import static io.curity.identityserver.plugin.OptInMFA.OptInMFAAuthenticationAction.CHOSEN_SECOND_FACTOR_ATTRIBUTE;
+import static io.curity.identityserver.plugin.OptInMFA.OptInMFAAuthenticationAction.IS_SECOND_FACTOR_CHOSEN_ATTRIBUTE;
+import static io.curity.identityserver.plugin.OptInMFA.OptInMFAAuthenticationAction.REMEMBER_CHOICE_COOKIE_NAME;
+import static java.util.Collections.EMPTY_MAP;
+import static se.curity.identityserver.sdk.web.ResponseModel.templateResponseModel;
+
+public final class OptInMFAuthenticationActionHandler implements ActionCompletionRequestHandler<Request>
+{
+    private static final Logger _logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final AuthenticatorDescriptorFactory _authenticatorDescriptorFactory;
+    private final SessionManager _sessionManager;
+    private final OptInMFAAuthenticationActionConfig _configuration;
+    private final ExceptionFactory _exceptionFactory;
+
+    public OptInMFAuthenticationActionHandler(
+            AuthenticatorDescriptorFactory factory,
+            SessionManager sessionManager,
+            OptInMFAAuthenticationActionConfig configuration,
+            ExceptionFactory exceptionFactory)
+    {
+        _authenticatorDescriptorFactory = factory;
+        _sessionManager = sessionManager;
+        _configuration = configuration;
+        _exceptionFactory = exceptionFactory;
+    }
+
+    @Override
+    public Optional<ActionCompletionResult> get(Request request, Response response)
+    {
+        Map<String, String> secondFactors = (Map<String, String>) _sessionManager.get(AVAILABLE_SECOND_FACTORS_ATTRIBUTE).getValue();
```

Comment:
There seems to be no reason to believe we will only receive this request when the session contains this attribute, so we should handel the case where it's null.

I would also verify the attribute has the right type:

```java
@Nullable Attribute secondFactors = _sessionManager.get(AVAILABLE_SECOND_FACTORS_ATTRIBUTE);
if (secondFactors != null)
{
    AttributeValue value = secondFactors.getAttributeValue();
    if (value instanceof MapAttributeValue)
    {
        Map<String, Object> secondFactors = ((MapAttributeValue) value).getValue();
        ...
    }
}
```

Quite a lot of work :/ but necessary.